--- a/compat/Kconfig	2017-01-15 20:20:30.153552885 +0100
+++ b/compat/Kconfig	2017-01-15 20:20:36.776900738 +0100
@@ -139,7 +139,7 @@ config BPAUTO_BUILD_WANT_DEV_COREDUMP
 config BPAUTO_RHASHTABLE
 	bool
 	# current API of rhashtable was introduced in version 4.1
-	depends on KERNEL_4_8
+	depends on KERNEL_4_9
 	# not very nice - but better than always having it
 	default y if MAC80211
 	#h-file linux/rhashtable.h
