--- a/arch/arm/boot/dts/qcom-ipq4019.dtsi
+++ b/arch/arm/boot/dts/qcom-ipq4019.dtsi
@@ -646,8 +646,12 @@
 			compatible = "qcom,ess-switch";
 			reg = <0xc000000 0x80000>;
 			switch_access_mode = "local bus";
-			resets = <&gcc ESS_RESET>;
-			reset-names = "ess_rst";
+			resets = <&gcc ESS_RESET>, <&gcc ESS_MAC1_CLK_DIS>,
+				 <&gcc ESS_MAC2_CLK_DIS>, <&gcc ESS_MAC3_CLK_DIS>,
+				 <&gcc ESS_MAC4_CLK_DIS>, <&gcc ESS_MAC5_CLK_DIS>;
+			reset-names = "ess_rst", "ess_mac1_clk_dis",
+				"ess_mac2_clk_dis", "ess_mac3_clk_dis",
+				"ess_mac4_clk_dis", "ess_mac5_clk_dis";
 			clocks = <&gcc GCC_ESS_CLK>;
 			clock-names = "ess_clk";
 			switch_cpu_bmp = <0x1>;
@@ -662,6 +666,8 @@
 			compatible = "qcom,ess-psgmii";
 			reg = <0x98000 0x800>;
 			psgmii_access_mode = "local bus";
+			resets = <&gcc ESS_PSGMII_ARES>;
+			reset-names = "psgmii_rst";
 			status = "disabled";
 		};
 
